typing/typedecl.ml

```diff
@@ -91,12 +93,37 @@ let update_type temp_env env id loc =
       with Ctype.Unify trace ->
         raise (Error(loc, Type_clash (env, trace)))
 
-(* Determine if a type is (an abbreviation for) the type "float" *)
 (* We use the Ctype.expand_head_opt version of expand_head to get access
    to the manifest type of private abbreviations. *)
+let rec get_unboxed_type_representation env ty fuel =
+  if fuel < 0 then None else
+  let ty = Ctype.repr (Ctype.expand_head_opt env ty) in
+  match ty.desc with
+  | Tconstr (p, args, _) ->
+    let tydecl = Env.find_type p env in
+    if Builtin_attributes.has_unboxed tydecl.type_attributes then begin
+      match tydecl.type_kind with
+      | Type_record ([{ld_type = ty2; _}], _)
+      | Type_variant [{cd_args = Cstr_tuple [ty2]; _}]
+      | Type_variant [{cd_args = Cstr_record [{ld_type = ty2; _}]; _}]
+      -> get_unboxed_type_representation env
+           (Ctype.apply env tydecl.type_params ty2 args) (fuel - 1)
+      | Type_abstract -> None
+          (* This case can occur when checking a recursive unboxed type
+             declaration. *)
+      | _ -> assert false (* only the above can be unboxed *)
+    end else
+      Some ty
+  | _ -> Some ty
+
+let get_unboxed_type_representation env ty =
+  get_unboxed_type_representation env ty 100000
```

Comment:
I feel uncomfortable with this fuel stuff, especially since when the fuel is exhausted (because of some kind of recursion), the function returns None, which is interpreted "positively" below in check_unboxed_gadt_arg.  Can we perhaps instead track visited "type paths" (even if this produces something a bit more restrictive for non-regular recursion)?